lib/hsa/mcwamp_hsa.cpp

```diff
@@ -1494,22 +1494,31 @@ class HSAQueue final : public KalmarQueue
     bool isEmpty() override {
         // Have to walk asyncOps since it can contain null pointers (if event is waited on and removed)
         // Also not all commands contain signals.
-        for (int i = 0; i < asyncOps.size(); ++i) {
-            if (asyncOps[i] != nullptr) {
-                auto &asyncOp = asyncOps[i];
+        bool oldestNoSignal=false;
+        for (int i = asyncOps.size()-1; i >= 0;  i--) {
```

Comment:
@bensander couldn't this be just be find_if with reverse iterators + the associated processing? Which is to say something along the lines of: 
```cpp
const auto oldest = find_if(
    asyncOps.crbegin(), asyncOps.crend(), [](... asyncOp) { return asyncOp != nullptr; });

if (oldest != asyncOps.crend()) { /* do the check for signal */ }
```
The loop is somewhat unnatural as is (it appears more customary to use while loops when counting down).

Reply:
yep, I think this can be refactored as you suggest, will submit a new one.  It started as a forward loop ; now that it runs backward all we really have to do is find the oldest op and see if it has a signal or not.